File: backend/src/dealers/credit_mix.rs
use crate::dealers::dealer::{CreditMix, CreditMixMonth, CreditMixShares};
use crate::dealers::queries::{DEALER_CREDIT_MIX, DEALER_CREDIT_MIX_PEER};
use crate::snowflake::{Row, SnowflakeClient};

pub async fn for_dealer(client: &SnowflakeClient, dealer_id: &str) -> Result<CreditMix, String> {
    let months_response = client
        .query_with_bindings(DEALER_CREDIT_MIX, &[dealer_id])
        .await?;
    let months = months(months_response.rows());

    let peer_response = client
        .query_with_bindings(DEALER_CREDIT_MIX_PEER, &[dealer_id, dealer_id])
        .await?;
    let peer_rows = peer_response.rows();
    let peer_row = peer_rows.first();

    let tier = peer_row.map(|row| field(row, "TIER")).unwrap_or_default();
    let peer = CreditMixShares {
        prime: peer_row.map(|r| number(r, "PRIME")).unwrap_or(0.0),
        near_prime: peer_row.map(|r| number(r, "NEAR_PRIME")).unwrap_or(0.0),
        subprime: peer_row.map(|r| number(r, "SUBPRIME")).unwrap_or(0.0),
    };

    let why = explain(&months, &peer, &tier);

    Ok(CreditMix {
        tier,
        months,
        peer,
        why,
    })
}

fn months(rows: Vec<Row>) -> Vec<CreditMixMonth> {
    let mut months: Vec<CreditMixMonth> = Vec::new();

    for row in rows {
        let period = field(&row, "PERIOD");
        let credit_tier = field(&row, "CREDIT_TIER");
        let count = count(&row);

        let month = match months.last_mut() {
            Some(month) if month.period == period => month,
            _ => {
                months.push(CreditMixMonth {
                    period,
                    prime: 0.0,
                    near_prime: 0.0,
                    subprime: 0.0,
                    volume: 0,
                });
                months.last_mut().unwrap()
            }
        };

        match credit_tier.as_str() {
            "Prime" => month.prime += count as f64,
            "Near_Prime" => month.near_prime += count as f64,
            "Subprime" => month.subprime += count as f64,
            _ => {}
        }
        month.volume += count;
    }

    for month in &mut months {
        if month.volume > 0 {
            let total = month.volume as f64;
            month.prime /= total;
            month.near_prime /= total;
            month.subprime /= total;
        }
    }

    months
}

fn explain(months: &[CreditMixMonth], peer: &CreditMixShares, tier: &str) -> String {
    let (Some(first), Some(last)) = (months.first(), months.last()) else {
        return String::new();
    };

    let prime_delta = (last.prime - first.prime) * 100.0;
    let sub_delta = (last.subprime - first.subprime) * 100.0;
    let prime_dir = if prime_delta < 0.0 { "fell" } else { "rose" };
    let sub_dir = if sub_delta >= 0.0 { "climbed" } else { "eased" };

    let trend = format!(
        "Prime share {} {:.0}%→{:.0}% since {}, while subprime {} {:.0}%→{:.0}%.",
        prime_dir,
        first.prime * 100.0,
        last.prime * 100.0,
        first.period,
        sub_dir,
        first.subprime * 100.0,
        last.subprime * 100.0,
    );

    let sub_gap = (last.subprime - peer.subprime) * 100.0;
    let peer_clause = if sub_gap.abs() < 1.0 {
        format!("Subprime mix is in line with the Tier {tier} peer average.")
    } else if sub_gap > 0.0 {
        format!(
            "Subprime now sits {:.0}pt above the Tier {} peer average of {:.0}%, a riskier pool that lifts decline rates.",
            sub_gap,
            tier,
            peer.subprime * 100.0,
        )
    } else {
        format!(
            "Subprime sits {:.0}pt below the Tier {} peer average of {:.0}%.",
            sub_gap.abs(),
            tier,
            peer.subprime * 100.0,
        )
    };

    format!("{trend} {peer_clause}")
}

fn field(row: &Row, column: &str) -> String {
    row.get(column).unwrap_or_default().to_string()
}

fn count(row: &Row) -> u64 {
    row.get("APPLICATION_COUNT")
        .and_then(|v| v.parse().ok())
        .unwrap_or(0)
}

fn number(row: &Row, column: &str) -> f64 {
    row.get(column).and_then(|v| v.parse().ok()).unwrap_or(0.0)
}

File: backend/app/dealers/credit_mix.py
from app.dealers.queries import DEALER_CREDIT_MIX, DEALER_CREDIT_MIX_PEER
from app.snowflake_client import Row, SnowflakeClient


def for_dealer(client: SnowflakeClient, dealer_id: str) -> dict:
    months_rows = client.query_with_bindings(DEALER_CREDIT_MIX, [dealer_id])
    months = _months(months_rows)

    peer_rows = client.query_with_bindings(DEALER_CREDIT_MIX_PEER, [dealer_id, dealer_id])
    peer_row = peer_rows[0] if peer_rows else None

    tier = _field(peer_row, "TIER") if peer_row else ""
    peer = {
        "prime": _number(peer_row, "PRIME") if peer_row else 0.0,
        "near_prime": _number(peer_row, "NEAR_PRIME") if peer_row else 0.0,
        "subprime": _number(peer_row, "SUBPRIME") if peer_row else 0.0,
    }

    why = _explain(months, peer, tier)

    return {
        "tier": tier,
        "months": months,
        "peer": peer,
        "why": why,
    }


def _months(rows: list[Row]) -> list[dict]:
    months: list[dict] = []

    for row in rows:
        period = _field(row, "PERIOD")
        credit_tier = _field(row, "CREDIT_TIER")
        count = _count(row)

        if months and months[-1]["period"] == period:
            month = months[-1]
        else:
            month = {
                "period": period,
                "prime": 0.0,
                "near_prime": 0.0,
                "subprime": 0.0,
                "volume": 0,
            }
            months.append(month)

        if credit_tier == "Prime":
            month["prime"] += count
        elif credit_tier == "Near_Prime":
            month["near_prime"] += count
        elif credit_tier == "Subprime":
            month["subprime"] += count
        month["volume"] += count

    for month in months:
        if month["volume"] > 0:
            total = float(month["volume"])
            month["prime"] /= total
            month["near_prime"] /= total
            month["subprime"] /= total

    return months


def _explain(months: list[dict], peer: dict, tier: str) -> str:
    if not months:
        return ""

    first, last = months[0], months[-1]

    prime_delta = (last["prime"] - first["prime"]) * 100
    sub_delta = (last["subprime"] - first["subprime"]) * 100
    prime_dir = "fell" if prime_delta < 0 else "rose"
    sub_dir = "climbed" if sub_delta >= 0 else "eased"

    trend = (
        f"Prime share {prime_dir} {first['prime'] * 100:.0f}%→{last['prime'] * 100:.0f}% "
        f"since {first['period']}, while subprime {sub_dir} "
        f"{first['subprime'] * 100:.0f}%→{last['subprime'] * 100:.0f}%."
    )

    sub_gap = (last["subprime"] - peer["subprime"]) * 100
    if abs(sub_gap) < 1.0:
        peer_clause = f"Subprime mix is in line with the Tier {tier} peer average."
    elif sub_gap > 0.0:
        peer_clause = (
            f"Subprime now sits {sub_gap:.0f}pt above the Tier {tier} peer average of "
            f"{peer['subprime'] * 100:.0f}%, a riskier pool that lifts decline rates."
        )
    else:
        peer_clause = (
            f"Subprime sits {abs(sub_gap):.0f}pt below the Tier {tier} peer average of "
            f"{peer['subprime'] * 100:.0f}%."
        )

    return f"{trend} {peer_clause}"


def _field(row: Row, column: str) -> str:
    return row.get(column) or ""


def _count(row: Row) -> int:
    value = row.get("APPLICATION_COUNT")
    try:
        return int(value)
    except (TypeError, ValueError):
        return 0


def _number(row: Row, column: str) -> float:
    value = row.get(column)
    try:
        return float(value)
    except (TypeError, ValueError):
        return 0.0
